engine/access/rpc/backend/connection_factory.go

```diff
@@ -61,16 +68,33 @@ func (cf *ConnectionFactoryImpl) createConnection(address string, timeout time.D
 	return conn, nil
 }
 
+func (cf *ConnectionFactoryImpl) retrieveConnection(address string) (*grpc.ClientConn, error) {
+	var conn *grpc.ClientConn
+	if res, ok := cf.ConnectionsCache.Get(address); ok {
+		conn = res.(*grpc.ClientConn)
+	}
+	if conn == nil || conn.GetState() != connectivity.Ready {
+		conn, err := cf.createConnection(address, cf.ExecutionNodeGRPCTimeout)
```

Comment:
otherwise, `conn` here is a different variable than the one returned on line 83
```suggestion
		var err error
		conn, err = cf.createConnection(address, cf.ExecutionNodeGRPCTimeout)
```